drivers/sensors/cirque_pinnacle_gestures.c

```diff
@@ -0,0 +1,205 @@
+        tap.timer = 0;
+    }
+
+    return mouse_report;
+}
+
+// extern this function for now
+void cirque_pinnacle_enable_tap(bool enable) {
+    features.tap_enable = enable;
+}
+
+// To set a trackpad exclusively as scroll wheel: outer_ring_pct = 100, trigger_px = 0, trigger_ang = 0
+static circular_scroll_context_t scroll = {.config = {.outer_ring_pct = 33, .trigger_px = 16, .trigger_ang = 9102 /* 50 degrees */, .wheel_clicks = 18}};
+
+static inline uint16_t atan2_16(int32_t dy, int32_t dx) {
+    if (dy == 0) {
+        if (dx >= 0)
+            return 0;
+        else
+            return 32768;
+    }
+
+    int32_t abs_y = dy > 0 ? dy : -dy;
+    int16_t a;
+
+    if (dx >= 0)
+        a = 8192 - (8192 * (dx - abs_y) / (dx + abs_y));
+    else
+        a = 24576 - (8192 * (dx + abs_y) / (abs_y - dx));
+
+    if (dy < 0) {
+        return -a; // negate if in quad III or IV
+    }
+    return a;
+}
+
+static circular_scroll_t circular_scroll(pinnacle_data_t touchData) {
+    circular_scroll_t report = {0, 0, false};
+    int8_t            x, y, wheel_clicks;
+    uint8_t           center = 256 / 2, mag;
+    int16_t           ang, dot, det, opposite_side, adjacent_side;
+    uint16_t          scale = cirque_pinnacle_get_scale();
+
+    if (touchData.zValue) {
+        // place origin at center of trackpad, treat coordinates as vectors
+        // scale to fixed int8_t size, angles are independent of resolution
+        if (scale) {
+            x = (int8_t)((int32_t)touchData.xValue * 256 / scale - center);
+            y = (int8_t)((int32_t)touchData.yValue * 256 / scale - center);
+        } else {
+            x = 0;
+            y = 0;
+        }
+
+        // check if first touch
+        if (!scroll.z) {
+            report.suppress_touch = false;
+            // check if touch falls within outer ring
+            mag = sqrt16(x * x + y * y);
+            if (mag * 100 / center >= 100 - scroll.config.outer_ring_pct) {
+                scroll.state = SCROLL_DETECTING;
+                scroll.x     = x;
+                scroll.y     = y;
+                scroll.mag   = mag;
+                // decide scroll axis:
+                //   vertical if started from righ half
+                //   horizontal if started from left half
+                // flipped for left-handed
+#if defined(POINTING_DEVICE_ROTATION_90)
+                scroll.axis = y < 0;
+#elif defined(POINTING_DEVICE_ROTATION_180)
+                scroll.axis = x > 0;
+#elif defined(POINTING_DEVICE_ROTATION_270)
+                scroll.axis = y > 0;
+#else
+                scroll.axis = x < 0;
+#endif
+            }
+        } else if (scroll.state == SCROLL_DETECTING) {
+            report.suppress_touch = true;
+            // already detecting scroll, check movement from touchdown location
+            mag = sqrt16((x - scroll.x) * (x - scroll.x) + (y - scroll.y) * (y - scroll.y));
+            if (mag >= scroll.config.trigger_px) {
+#define ABS(a) (a > 0 ? a : -a)
```

Comment:
any tradeoff to consider when moving this to a static-inline function? mixing both "styles" with defines and inline function feels a bit inconsistent (although ABS is very short... so maybe ok?)

also: is there no ABS in any of the available math libs?

Reply:
Thanks for bringing this up! Apparently I'd spend too long in an environment where abs() is overloaded with the version for doubles, and completely forgot the stdlib integer version even existed.